packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -163,14 +200,37 @@ abstract class Route<T> {
   @mustCallSuper
   void didChangePrevious(Route<dynamic> previousRoute) { }
 
-  /// The route was popped or is otherwise being removed somewhat gracefully.
+  /// Called whenever the internal state of the route has changed.
   ///
-  /// This is called by [didPop] and in response to [Navigator.pushReplacement].
+  /// This should be called whenever [willHandlePopInternally], [didPop],
+  /// [offstage], or other internal state of the route changes value. It is used
+  /// by [ModalRoute], for example, to report the new information via its
+  /// inherited widget to any children of the route.
+  ///
+  /// See also:
+  ///
+  ///  * [changedExternalState], which is called when the [Navigator] rebuilds.
   @protected
   @mustCallSuper
-  void didComplete(T result) {
-    _popCompleter.complete(result);
-  }
+  void changedInternalState() { }
+
+  /// Called whenever the [Navigator] has its widget rebuilt, to indicate that
+  /// the route may wish to rebuild as well.
+  ///
+  /// This is called by the [Navigator] whenever the [NavigatorState]'s
+  /// widget changes, for example because the [MaterialApp] has been rebuilt.
```

Comment:
Not entirely clear for me what "[NavigatorState]'s widget" is. Did you mean the output of [NavigatorState.build]?

Reply:
Is it clearer if I say: "[NavigatorState]'s [widget]" ?

Reply:
Ah that's what you meant. Yes. 

Or could you say `This is called whenever the [Navigator] is no longer equal to the previous [Navigator] in the widget tree` if I understood correctly?

Reply:
I'd rather avoid the term "widget tree" here since what really matters is that the widgets associated with the elements in the element tree have changed identity but saying that precisely won't help anyone except people who already understand it...